### PR TITLE
feat: merge allowlist and erc20 spend limit modules

### PR DESCRIPTION
**Pull request by fangting-alchemy**
## Motivation
ERC20Token spend limit cannot be used without Allowlist module.

## Solution
Merge ERC20 spend limit with allowlist modules.

### CODE REVIEW

File: src/modules/permissions/AllowlistModule.sol [hunk 1]
```diff
@@ -25,22 +27,47 @@ import {BaseModule} from "../../modules/BaseModule.sol";
...
 ///       - if sepecific address + sepecific selector, pass
 ///       - revert all other cases
-contract AllowlistModule is IValidationHookModule, BaseModule {
+///    - Uninstallation will not remove all state for account. Account must pass all data to remove all rules for
+/// an validation entity.
+/// If specific address is set, a ERC20TokenSpendLimit might be set for it. If so, the following features and
+/// restrictions apply:
+///    - The execution hooks to be associated with validations should be used.
+///    - For any validation function with this execution hook installed, only token contracts with a set limit will
+/// be
```
Comment (Zer0dot) [inline src/modules/permissions/AllowlistModule.sol]: Maybe some minor formatting could be done here

Comment (Zer0dot) [reply src/modules/permissions/AllowlistModule.sol]: (I know how annoying the automatic line length adjustments from the formatter can be)

File: src/modules/permissions/AllowlistModule.sol [hunk 2]
```diff
@@ -73,6 +105,29 @@ contract AllowlistModule is IValidationHookModule, BaseModule {
...
+        if (selector == IModularAccount.execute.selector) {
+            // when calling execute or ERC20 functions directly
+            (address token,, bytes memory innerCalldata) = abi.decode(callData, (address, uint256, bytes));
+            _decrementLimitIfApplies(entityId, token, innerCalldata);
+        } else if (selector == IModularAccount.executeBatch.selector) {
+            Call[] memory calls = abi.decode(callData, (Call[]));
+            for (uint256 i = 0; i < calls.length; i++) {
+                _decrementLimitIfApplies(entityId, calls[i].target, calls[i].data);
+            }
+        } else {
```
Comment (Zer0dot) [inline src/modules/permissions/AllowlistModule.sol:124]: So basically, you should never install the validation-associated execution hook on anything but exec/execBatch right?

File: src/modules/permissions/AllowlistModule.sol [hunk 3]
```diff
@@ -204,7 +280,36 @@ contract AllowlistModule is IValidationHookModule, BaseModule {
...
+        uint256 spendLimit = erc20SpendLimits[entityId][token][msg.sender];
+
+        if (innerCalldata.length < 68) {
+            revert InvalidCalldataLength();
+        }
+
+        bytes4 selector;
+        uint256 spend;
+        assembly ("memory-safe") {
+            selector := mload(add(innerCalldata, 32)) // 0:32 is arr len, 32:36 is selector
```
Comment (Zer0dot) [inline src/modules/permissions/AllowlistModule.sol]:
```suggestion
            selector := mload(add(innerCalldata, 0x20)) // 0x00:0x20 is arr len, 0x20:0x24 is selector
```
Just for consistency when working with memory I think we opt to use hexadecimal offsets, but this is a nitpick.

File: src/modules/permissions/AllowlistModule.sol [hunk 4]
```diff
@@ -204,7 +280,36 @@ contract AllowlistModule is IValidationHookModule, BaseModule {
...
+            spend := mload(add(innerCalldata, 68)) // 36:68 is recipient, 68:100 is spend
```
Comment (Zer0dot) [inline src/modules/permissions/AllowlistModule.sol]:
```suggestion
            spend := mload(add(innerCalldata, 0x44)) // 0x24:0x44 is recipient, 0x44:0x64 is spend
```
Same as above

File: test/modules/AllowlistERC20TokenLimit.t.sol [hunk 5]
```diff
@@ -13,50 +13,64 @@ import {HookConfigLib} from "../../src/libraries/HookConfigLib.sol";
...
 import {ValidationConfigLib} from "../../src/libraries/ValidationConfigLib.sol";
-import {ERC20TokenLimitModule} from "../../src/modules/permissions/ERC20TokenLimitModule.sol";
+import {AllowlistModule} from "../../src/modules/permissions/AllowlistModule.sol";
 
 import {MockModule} from "../mocks/modules/MockModule.sol";
 import {AccountTestBase} from "../utils/AccountTestBase.sol";
 import {CODELESS_ADDRESS} from "../utils/TestConstants.sol";
 
-contract ERC20TokenLimitModuleTest is AccountTestBase {
+contract AllowlistERC20TokenLimitTest is AccountTestBase {
```
Comment (Zer0dot) [inline test/modules/AllowlistERC20TokenLimit.t.sol:23]: Is there a reason we're not combining this test with the AllowlistModuleTest contract?

Also, can we add `withSMATest()` modifiers to the tests? (Except ones which only touch the module and not the account)

**Comment by Zer0dot** [comment]
No further comments beyond what I've got, happy to approve once addressed!

File: src/modules/permissions/AllowlistModule.sol (hunk 2 above)
Comment (fangting-alchemy) [reply src/modules/permissions/AllowlistModule.sol:124]: Depends on what you are trying to check on. In our case, `executeUserOp` calls self, so checking those two seems enough.

But users may install other execution functions that might do something. They can have a validation-associated hook on the validator that has access to that execution function.

File: test/modules/AllowlistERC20TokenLimit.t.sol (hunk 5 above)
Comment (fangting-alchemy) [reply test/modules/AllowlistERC20TokenLimit.t.sol:23]: I thought it was better organized this way to make sure we test both features of the module.

All tests have `withSMATest` modifier already.

File: src/modules/permissions/AllowlistModule.sol (hunk 2 above)
Comment (Zer0dot) [reply src/modules/permissions/AllowlistModule.sol:124]: yeah but you shouldn't install *this* one on anything but those two or you will revert yeah?

Comment (fangting-alchemy) [reply src/modules/permissions/AllowlistModule.sol:124]: Sry, miss read. For this hook, yes.

**Review by Zer0dot — APPROVED** [review]

**Pull request merged**